pkg/cli/updater/resolver/resolver_test.go

```diff
@@ -58,42 +38,167 @@ func TestResolve(t *testing.T) {
+			name: "should return a version between constraints",
+			c:    Criteria{Constraints: []string{">0.9.0", "<1.0.0"}},
+			versions: map[string][]Version{
+				StableChannel: {
+					newTestingVersion("v1.0.0"),
+					newTestingVersion("v0.9.1"),
+					newTestingVersion("v0.9.0"),
+				},
+			},
+			want: newTestingVersion("v0.9.1"),
+		},
+		{
+			name: "should satisfy a version constraint when there are multiple channels",
+			c:    Criteria{Channel: "rc", Constraints: []string{">=0.9.0"}},
+			versions: map[string][]Version{
+				StableChannel: {
+					newTestingVersion("v0.9.1"),
+				},
+				"rc": {
+					// this older than v0.9.1
+					newTestingVersion("v0.9.1-rc.1"),
+				},
+			},
+			want: newTestingVersion("v0.9.1"),
+		},
+		{
+			name: "should satisfy a version constraint when there are multiple channels with a pre-release being higher",
+			c:    Criteria{Channel: "rc", Constraints: []string{">=0.9.0"}},
+			versions: map[string][]Version{
+				StableChannel: {
+					newTestingVersion("v0.9.0"),
+				},
+				"rc": {
+					newTestingVersion("v0.9.1-rc.1"),
+				},
+			},
+			want: newTestingVersion("v0.9.1-rc.1"),
+		},
+		{
+			name: "should return a version outside of the channel when a constraint is provided asking for it",
+			c:    Criteria{Constraints: []string{"~0.9.1-rc"}},
+			versions: map[string][]Version{
+				StableChannel: {
+					newTestingVersion("v0.9.0"),
+				},
+				"rc": {
+					newTestingVersion("v0.9.1-rc.1"),
+				},
+			},
+			want: newTestingVersion("v0.9.1-rc.1"),
+		},
+		{
+			name: "should only opt-into a channel when a constraint is provided or in the channel",
+			c:    Criteria{Constraints: []string{">=0.9.1-rc.1"}},
+			versions: map[string][]Version{
+				StableChannel: {
+					newTestingVersion("v0.9.0"),
+				},
+				"rc": {
+					newTestingVersion("v0.9.1-rc.1"),
+				},
+				"beta": {
+					// should be ignored
+					newTestingVersion("v0.9.2-beta.1"),
+				},
+			},
+			want: newTestingVersion("v0.9.1-rc.1"),
+		},
+		{
+			name: "should support channel with constraint wanting another channel",
+			c:    Criteria{Channel: "rc", Constraints: []string{">=0.9.1-beta"}},
+			versions: map[string][]Version{
+				StableChannel: {
+					newTestingVersion("v0.9.0"),
+				},
+				"rc": {
+					newTestingVersion("v0.9.1-rc.1"),
+				},
+				"beta": {
+					newTestingVersion("v0.9.2-beta.1"),
+				},
+			},
+			// should return beta because module asked for it
+			want: newTestingVersion("v0.9.2-beta.1"),
+		},
+		{
+			name: "should support channel being gtr with constraint wanting another channel",
+			c:    Criteria{Channel: "beta", Constraints: []string{">=0.9.1-rc"}},
+			versions: map[string][]Version{
+				StableChannel: {
+					newTestingVersion("v0.9.0"),
+				},
+				"rc": {
+					newTestingVersion("v0.9.1-rc.1"),
+				},
+				"beta": {
+					newTestingVersion("v0.9.2-beta.1"),
+				},
+			},
+			// should return beta because channel asked for it
+			want: newTestingVersion("v0.9.2-beta.1"),
```

Comment:
This looks great! Just to cover all test cases, we should add a test where we compare two rc versions (e.g. `v0.9.1-rc.1` and `v0.9.1-rc.2`).